src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/AbstractTypesShouldNotHaveConstructors.Fixer.cs

```diff
@@ -45,7 +45,27 @@ private static async Task<Document> ChangeAccessibilityCodeFix(Document document
         {
             SemanticModel model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
             var classSymbol = (INamedTypeSymbol)model.GetDeclaredSymbol(nodeToFix, cancellationToken);
-            System.Collections.Generic.List<SyntaxNode> instanceConstructors = classSymbol.InstanceConstructors.Where(t => t.DeclaredAccessibility == Accessibility.Public).Select(t => GetDeclaration(t)).Where(d => d != null).ToList();
```

Comment:
In general, we do not restrict linq usage in analyzers/fixers unless we have performance traces showing a noticeable difference. I understand avoiding linq in hot paths, but I think readability is more important otherwise. Personally, I prefer the readability of the original code, and if we were to make changes like this, we would need to move the implementation details (loop being added) into a separate local function.

Tagging @dotnet/roslyn-analysis for other opinions here.

Reply:
I wonder what @dotnet/roslyn has to say about this.

I remember someone saying Roslyn had cut back on the use of LINQ because it was hitting performance.

And it's not just because of these analyzers, It's because analyzers are citizens inside the compiler society. Any better citizen makes a better society that can be ruined by just one citizen.

Reply:
This is not true. Linq is used almost everywhere in Roslyn from Workspace layer upwards. Compiler team prefers to avoid linq, but that is more a team preference. I agree with @drewnoakes's question: PRs claiming to improve performance should be driven via performance analysis and show how the changes are improving performance. I would be fine with accepting these changes in hot paths, but I am not sure banning linq everywhere is a good approach.

Reply:
It also depends on the expectations. VSMac currently runs all Analyzers in the IDE process, thus gets GC hits on every keypress due to analyzers running on every keypress.

Our GC configuration on Mono currently cannot handle lots of small allocations well, thus we get GC pauses hitting the user.

Although, I agree on getting numbers first. Trace driven optimizing is the way to to most of the way.

We don't ship these analyzers by default, but they can be enabled in various projects. Just wanted to get that into consideration too. Most of the optimization work done recently in VSMac was to reduce gen0 allocations.

Reply:
Our ideal scenario is to use LINQ when things are actually lazy, not when we're working with data computed eagerly